pkg/clangtool/clangtool.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package clangtool
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/google/syzkaller/pkg/declextract"
+	"github.com/google/syzkaller/pkg/osutil"
+)
+
+type Config struct {
+	ToolBin    string
+	KernelSrc  string
+	KernelObj  string
+	CacheDir   string
+	ReuseCache bool
+}
+
+// Run runs the clang tool on all files in the compilation database
+// in the kernel build dir and returns combined output for all files.
+// It always caches results, and optionally reuses previously cached results.
+func Run(cfg *Config) (*declextract.Output, error) {
+	dbFile := filepath.Join(cfg.KernelObj, "compile_commands.json")
+	cmds, err := loadCompileCommands(dbFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load compile commands: %w", err)
+	}
+
+	type result struct {
+		out *declextract.Output
+		err error
+	}
+	results := make(chan *result, 10)
```

Comment:
Is there a reason for it to be `len(cmds)`?
Using such size is usually a signal that it's required to prevent deadlocks, but it's not the case here, so it will be confusing.